In [86]:
!./neofetch

            .-/+oossssoo+/-.
        `:+ssssssssssssssssss+:`
      -+ssssssssssssssssssyyssss+-
    .ossssssssssssssssssdMMMNysssso.
   /ssssssssssshdmmNNmmyNMMMMhssssss/
  +ssssssssshmydMMMMMMMNddddyssssssss+
 /sssssssshNMMMyhhyyyyhmNMMMNhssssssss/
.ssssssssdMMMNhsssssssssshNMMMdssssssss.
+sssshhhyNMMNyssssssssssssyNMMMysssssss+
ossyNMMMNyMMhsssssssssssssshmmmhssssssso
ossyNMMMNyMMhsssssssssssssshmmmhssssssso
+sssshhhyNMMNyssssssssssssyNMMMysssssss+
.ssssssssdMMMNhsssssssssshNMMMdssssssss.
 /sssssssshNMMMyhhyyyyhdNMMMNhssssssss/
  +sssssssssdmydMMMMMMMMddddyssssssss+
   /ssssssssssshdmNNNNmyNMMMMhssssss/
    .ossssssssssssssssssdMMMNysssso.
      -+sssssssssssssssssyyyssss+-
        `:+ssssssssssssssssss+:`
            .-/+oossssoo+/-.
root@1041049adf5c 
----------------- 
OS: Ubuntu 16.04.6 LTS x86_64 
Host: HVM domU 4.2.amazon 
Kernel: 4.4.0-1102-aws 
Uptime: 37 mins 
Packages: 237 (dpkg) 
Shell: bash 4.3.48 
CPU: Intel Xeon E5-2686 v4 (32) @ 3.000GHz 
Memory: 4536MiB / 245854MiB 


In [128]:
!nvidia-smi

Mon Nov  9 17:42:27 2020       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 440.33.01    Driver Version: 440.33.01    CUDA Version: 10.2     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla V100-SXM2...  Off  | 00000000:00:1B.0 Off |                    0 |
| N/A   58C    P0    58W / 300W |   4300MiB / 16160MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   1  Tesla V100-SXM2...  Off  | 00000000:00:1C.0 Off |                    0 |
| N/A   49C    P0    37W / 300W |     11MiB / 16160MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   2  T

In [88]:
!cat /proc/cpuinfo

processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 79
model name	: Intel(R) Xeon(R) CPU E5-2686 v4 @ 2.30GHz
stepping	: 1
microcode	: 0xb000038
cpu MHz		: 2432.339
cache size	: 46080 KB
physical id	: 0
siblings	: 32
core id		: 0
cpu cores	: 16
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 13
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc aperfmperf pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single kaiser fsgsbase bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx xsaveopt
bugs		: cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf mds swapgs taa itlb_multihit
bogomips	: 4600.10
clflush size	: 64
cache_alignment	: 64
address sizes	: 46 bits physical, 48 bits virtual
power manag

In [31]:
import time, torch

bas = time.time()

a = torch.rand(10000, 10000, device=torch.device("cuda"))
b = torch.rand(10000, 10000, device=torch.device("cuda"))
c = torch.rand(10000, 10000, device=torch.device("cuda"))
d = torch.rand(10000, 10000, device=torch.device("cuda"))


for i in range(0,1000):
  y = a@b
  z = c@d

  x = y@z

son = time.time()

In [32]:
x

tensor([[6.3141e+10, 6.2336e+10, 6.2886e+10,  ..., 6.2182e+10, 6.1760e+10,
         6.2553e+10],
        [6.2995e+10, 6.2191e+10, 6.2740e+10,  ..., 6.2037e+10, 6.1616e+10,
         6.2408e+10],
        [6.3517e+10, 6.2707e+10, 6.3260e+10,  ..., 6.2552e+10, 6.2127e+10,
         6.2926e+10],
        ...,
        [6.2748e+10, 6.1948e+10, 6.2494e+10,  ..., 6.1794e+10, 6.1375e+10,
         6.2164e+10],
        [6.3345e+10, 6.2536e+10, 6.3088e+10,  ..., 6.2382e+10, 6.1958e+10,
         6.2755e+10],
        [6.3316e+10, 6.2508e+10, 6.3060e+10,  ..., 6.2354e+10, 6.1930e+10,
         6.2727e+10]], device='cuda:0')

In [33]:
"1.test result (second) : " + str(son-bas)

'1.test result (second) : 291.4778277873993'

In [2]:
!nvcc --version

nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2019 NVIDIA Corporation
Built on Wed_Oct_23_19:24:38_PDT_2019
Cuda compilation tools, release 10.2, V10.2.89


In [3]:
!pip install git+git://github.com/andreinechaev/nvcc4jupyter.git
%load_ext nvcc_plugin

  Cloning git://github.com/andreinechaev/nvcc4jupyter.git to /tmp/pip-req-build-rea45rtz
  Created wheel for NVCCPlugin: filename=NVCCPlugin-0.0.2-py3-none-any.whl size=4307 sha256=00bf7c94f8023a946bbf809ec7745c9242ef018871f80a107babc444a9d37b71
  Stored in directory: /tmp/pip-ephem-wheel-cache-b_f3g77n/wheels/c5/2b/c0/87008e795a14bbcdfc7c846a00d06981916331eb980b6c8bdf
Successfully built NVCCPlugin
created output directory at /dli/task/src
Out bin /dli/task/result.out


In [4]:
!wget https://ahmetfurkandemir.s3.amazonaws.com/kernel.cu

--2020-11-09 18:19:00--  https://ahmetfurkandemir.s3.amazonaws.com/kernel.cu
Resolving ahmetfurkandemir.s3.amazonaws.com (ahmetfurkandemir.s3.amazonaws.com)... 52.216.206.139
Connecting to ahmetfurkandemir.s3.amazonaws.com (ahmetfurkandemir.s3.amazonaws.com)|52.216.206.139|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1155 (1.1K) [application/cu-seeme]
Saving to: ‘kernel.cu’

kernel.cu           100%[===================>]   1.13K  --.-KB/s    in 0s      

2020-11-09 18:19:00 (93.0 MB/s) - ‘kernel.cu’ saved [1155/1155]



In [5]:
!cat kernel.cu

#include <math.h>
#include <iostream>
#include "cuda_runtime.h"
#include "kernel.h"
#include <stdlib.h>

using namespace std;

__global__ void matrixMultiplicationKernel(float* A, float* B, float* C, int N) {

    int ROW = blockIdx.y*blockDim.y+threadIdx.y;
    int COL = blockIdx.x*blockDim.x+threadIdx.x;

    float tmpSum = 0;

    if (ROW < N && COL < N) {
        // each thread computes one element of the block sub-matrix
        for (int i = 0; i < N; i++) {
            tmpSum += A[ROW * N + i] * B[i * N + COL];
        }
    }
    C[ROW * N + COL] = tmpSum;
}


void matrixMultiplication(float *A, float *B, float *C, int N){

    // declare the number of blocks per grid and the number of threads per block
    // use 1 to 512 threads per block
    dim3 threadsPerBlock(N, N);
    dim3 blocksPerGrid(1, 1);
        if (N*N > 512){
            threadsPerBlock.x = 512;
            threadsPerBlock.y = 512;
            blocksPerGrid.x = ceil(double(N)/double(threadsPerBlock.x));
         

In [6]:
!wget https://ahmetfurkandemir.s3.amazonaws.com/dev_array.h

--2020-11-09 18:19:01--  https://ahmetfurkandemir.s3.amazonaws.com/dev_array.h
Resolving ahmetfurkandemir.s3.amazonaws.com (ahmetfurkandemir.s3.amazonaws.com)... 52.216.206.139
Connecting to ahmetfurkandemir.s3.amazonaws.com (ahmetfurkandemir.s3.amazonaws.com)|52.216.206.139|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1997 (2.0K) [text/x-chdr]
Saving to: ‘dev_array.h’

dev_array.h         100%[===================>]   1.95K  --.-KB/s    in 0s      

2020-11-09 18:19:01 (149 MB/s) - ‘dev_array.h’ saved [1997/1997]



In [7]:
!cat dev_array.h

#ifndef _DEV_ARRAY_H_
#define _DEV_ARRAY_H_

#include <stdexcept>
#include <algorithm>
#include <cuda_runtime.h>

template <class T>
class dev_array
{
// public functions
public:
    explicit dev_array()
        : start_(0),
          end_(0)
    {}

    // constructor
    explicit dev_array(size_t size)
    {
        allocate(size);
    }
    // destructor
    ~dev_array()
    {
        free();
    }

    // resize the vector
    void resize(size_t size)
    {
        free();
        allocate(size);
    }

    // get the size of the array
    size_t getSize() const
    {
        return end_ - start_;
    }

    // get data
    const T* getData() const
    {
        return start_;
    }

    T* getData()
    {
        return start_;
    }

    // set
    void set(const T* src, size_t size)
    {
        size_t min = std::min(size, getSize());
        cudaError_t result = cudaMemcpy(start_, src, min * sizeof(T), cudaMemcpyHostToDevice);
        if (result != cudaSuccess)
        {
     

In [8]:
!wget https://ahmetfurkandemir.s3.amazonaws.com/kernel.h

--2020-11-09 18:19:03--  https://ahmetfurkandemir.s3.amazonaws.com/kernel.h
Resolving ahmetfurkandemir.s3.amazonaws.com (ahmetfurkandemir.s3.amazonaws.com)... 52.217.100.244
Connecting to ahmetfurkandemir.s3.amazonaws.com (ahmetfurkandemir.s3.amazonaws.com)|52.217.100.244|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 112 [text/x-chdr]
Saving to: ‘kernel.h’

kernel.h            100%[===================>]     112  --.-KB/s    in 0s      

2020-11-09 18:19:03 (4.10 MB/s) - ‘kernel.h’ saved [112/112]



In [9]:
!cat kernel.h

#ifndef KERNEL_CUH_
#define KERNEL_CUH_

void matrixMultiplication(float *A, float *B, float *C, int N);

#endif

In [10]:
!wget https://ahmetfurkandemir.s3.amazonaws.com/matrixmul.cu

--2020-11-09 18:19:06--  https://ahmetfurkandemir.s3.amazonaws.com/matrixmul.cu
Resolving ahmetfurkandemir.s3.amazonaws.com (ahmetfurkandemir.s3.amazonaws.com)... 52.217.36.140
Connecting to ahmetfurkandemir.s3.amazonaws.com (ahmetfurkandemir.s3.amazonaws.com)|52.217.36.140|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1060 (1.0K) [application/cu-seeme]
Saving to: ‘matrixmul.cu’

matrixmul.cu        100%[===================>]   1.04K  --.-KB/s    in 0s      

2020-11-09 18:19:06 (79.1 MB/s) - ‘matrixmul.cu’ saved [1060/1060]



In [11]:
!cat matrixmul.cu

#include <iostream>
#include <vector>
#include <stdlib.h>
#include <time.h>
#include <cuda_runtime.h>
#include "kernel.h"
#include "kernel.cu"
#include "dev_array.h"
#include <math.h>
#include <stdio.h>

using namespace std;

int main()
{
    // Perform matrix multiplication C = A*B
    // where A, B and C are NxN matrices
    int N = 10000;
    int SIZE = N*N;

    // Allocate memory on the host
    vector<float> h_A(SIZE);
    vector<float> h_B(SIZE);
    vector<float> h_C(SIZE);

    // Initialize matrices on the host
    for (int i=0; i<N; i++){
        for (int j=0; j<N; j++){
            h_A[i*N+j] = sin(i);
            h_B[i*N+j] = cos(j);
        }
    }

    // Allocate memory on the device
    dev_array<float> d_A(SIZE);
    dev_array<float> d_B(SIZE);
    dev_array<float> d_C(SIZE);

    d_A.set(&h_A[0], SIZE);
    d_B.set(&h_B[0], SIZE);

    matrixMultiplication(d_A.getData(), d_B.getData(), d_C.getData(), N);
    cudaDeviceSynchronize();

    d_C.get(&h_C[0], SIZE);
    c

In [15]:
# compilation test
bas = time.time()
!nvcc matrixmul.cu
son = time.time()

In [16]:
"2. test - compilation test result : " + str(son-bas)

'2. test - compilation test result : 1.413379192352295'

In [19]:
# run the compiled file, test
bas = time.time()
!./a.out
son = time.time()

END

In [20]:
"2. test - time to run the compiled file : "+str(son-bas)

'2. test - time to run the compiled file : 9.453376293182373'

**1. Test Result : 291.4778277873993 (second)**

**2. Test Reult - Compile : 1.413379192352295 (second)**

**2. Test Reult - run file : 9.453376293182373 (second)**